ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -3,52 +3,340 @@
+        assertEquals(
+          "Paint agreement parameters",
+          agreement.getCreateArguments,
+          Record(
+            recordId = Some(Tag.unwrap(PaintAgree.id)),
+            fields = Seq(
+              RecordField("painter", Some(Value(Value.Sum.Party(Tag.unwrap(painter))))),
+              RecordField("houseOwner", Some(Value(Value.Sum.Party(Tag.unwrap(houseOwner)))))
+            )
+          )
+        )
+      }
+    }
+
+  private[this] val successfulPaintCounterOffer =
+    LedgerTest("SemanticPaintCounterOffer", "Conduct the paint counter-offer worflow successfully") {
+      ledger =>
+        for {
+          Vector(bank, houseOwner, painter) <- ledger.allocateParties(3)
+          iou <- ledger.create(bank, Iou(bank, houseOwner, onePound))
+          offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, twoPounds))
+          counter <- ledger.exerciseAndGetContract[PaintCounterOffer](
+            houseOwner,
+            offer.exercisePaintOffer_Counter(_, iou))
+          tree <- ledger.exercise(painter, counter.exercisePaintCounterOffer_Accept)
+        } yield {
+          val agreement = assertSingleton(
+            "SemanticPaintCounterOffer",
+            createdEvents(tree).filter(_.getTemplateId == Tag.unwrap(PaintAgree.id)))
+          assertEquals(
+            "Paint agreement parameters",
+            agreement.getCreateArguments,
+            Record(
+              recordId = Some(Tag.unwrap(PaintAgree.id)),
+              fields = Seq(
+                RecordField("painter", Some(Value(Value.Sum.Party(Tag.unwrap(painter))))),
+                RecordField("houseOwner", Some(Value(Value.Sum.Party(Tag.unwrap(houseOwner)))))
+              )
+            )
+          )
+        }
+    }
+
+  private[this] val partialSignatories =
+    LedgerTest(
+      "SemanticPartialSignatories",
+      "A signatory should not be able to create a contract on behalf of two parties") { ledger =>
+      for {
+        Vector(houseOwner, painter) <- ledger.allocateParties(2)
+        failure <- ledger.create(houseOwner, PaintAgree(painter, houseOwner)).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "requires authorizers")
+      }
+    }
+
+  private[this] val acceptOnBehalf =
+    LedgerTest(
+      "SemanticAcceptOnBehalf",
+      "It should not be possible to exercise a choice without the consent of the controller") {
+      ledger =>
+        for {
+          Vector(bank, houseOwner, painter) <- ledger.allocateParties(3)
+          iou <- ledger.create(bank, Iou(bank, houseOwner, onePound))
+          offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+          failure <- ledger.exercise(painter, offer.exercisePaintOffer_Accept(_, iou)).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "couldn't find contract")
```

Comment:
The error code looks wrong here. We'd want an authorization failure and not a privacy failure. The problem is that this `Iou` is not visible to the painter.

An easy way out is to add the option to `Reject` the offer for the houseOwner and try to trigger that rejection as the painter. In that case no Iou needs to be created.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2718/commits/c0e8912d2ea892eaa3ff0f3e22c2abdc39150d6f

The cause was that the painter is unable to see the `iou`. I worked around the problem by giving the `iou` to the painter.

This is a very good catch.

![giphy](https://user-images.githubusercontent.com/43749967/64102478-5eadf200-cd70-11e9-8271-22be88de048a.gif)

This actually highlights a possible problem in how we do the validation, as @gerolf-da just told me offline.

I'd say we can (for now) limit ourselves to make sure the test passes but we can easily add another test that checks that the validation is performed in the correct order.